### PR TITLE
[Enhance] support eta_min_ratio in CosineAnnealingParamScheduler

### PR DESCRIPTION
**Pull request by cir7**
## Motivation
[Enhance] support eta_min_ratio in CosineAnnealingParamScheduler
[doc] fix a typo in CosineRestartParamScheduler's docstring

### CODE REVIEW

File: mmengine/optim/scheduler/param_scheduler.py [hunk 1]
```diff
@@ -599,7 +599,10 @@ class CosineAnnealingParamScheduler(_ParamScheduler):
             ``lr``, ``momentum``.
         T_max (int, optional): Maximum number of iterations. If not specified,
             use ``end - begin``. Defaults to None.
-        eta_min (float): Minimum parameter value. Defaults to 0.
+        eta_min (float): Minimum parameter value. Defaults to None.
```
Comment (RangiLyu) [inline mmengine/optim/scheduler/param_scheduler.py:602]:
```suggestion
        eta_min (float, optional): Minimum parameter value. Defaults to None.
```

**Review by RangiLyu — COMMENTED** [review]
Please also add some unit tests for eta_min_ratio.

**Review by RangiLyu — APPROVED** [review]

File: mmengine/optim/scheduler/param_scheduler.py [hunk 2]
```diff
@@ -599,7 +599,10 @@ class CosineAnnealingParamScheduler(_ParamScheduler):
             ``lr``, ``momentum``.
         T_max (int, optional): Maximum number of iterations. If not specified,
             use ``end - begin``. Defaults to None.
-        eta_min (float): Minimum parameter value. Defaults to 0.
+        eta_min (float, optional): Minimum parameter value. Defaults to None.
```
Comment (HAOCHENYE) [inline mmengine/optim/scheduler/param_scheduler.py:602]:
```suggestion
        eta_min (float, optional): Minimum parameter value.
```

File: mmengine/optim/scheduler/param_scheduler.py [hunk 3]
```diff
@@ -599,7 +599,10 @@ class CosineAnnealingParamScheduler(_ParamScheduler):
...
+        eta_min_ratio (float, optional): The ratio of minimum parameter value
+            to the base parameter value. Either `eta_min` or `eta_min_ratio`
+            should be specified. Default: None.
```
Comment (HAOCHENYE) [inline mmengine/optim/scheduler/param_scheduler.py:605]:
```suggestion
        eta_min_ratio (float, optional): The ratio of the minimum parameter value
            to the base parameter value. Either `eta_min` or `eta_min_ratio`
            should be specified. 
```

File: mmengine/optim/scheduler/param_scheduler.py [hunk 4]
```diff
@@ -619,14 +622,17 @@ def __init__(self,
...
                  T_max: Optional[int] = None,
-                 eta_min: float = 0.,
+                 eta_min: Optional[float] = None,
+                 eta_min_ratio: Optional[float] = None,
                  begin: int = 0,
                  end: int = INF,
                  last_step: int = -1,
                  by_epoch: bool = True,
                  verbose: bool = False):
+        assert (eta_min is None) ^ (eta_min_ratio is None)
```
Comment (HAOCHENYE) [inline mmengine/optim/scheduler/param_scheduler.py:632]: Raise an error information

File: mmengine/optim/scheduler/param_scheduler.py [hunk 5]
```diff
@@ -1130,10 +1144,10 @@ class CosineRestartParamScheduler(_ParamScheduler):
         periods (list[int]): Periods for each cosine anneling cycle.
         restart_weights (list[float]): Restart weights at each
             restart iteration. Defaults to [1].
-        eta_min (float): Minimum parameter value at the end of scheduling.
-            Defaults to None.
+        eta_min (float, optional): Minimum parameter value at the end of
+            scheduling. Defaults to None.
```
Comment (HAOCHENYE) [inline mmengine/optim/scheduler/param_scheduler.py:1148]:
```suggestion
        eta_min (float, optional): Minimum parameter value at the end of
            scheduling.
```

File: mmengine/optim/scheduler/param_scheduler.py [hunk 6]
```diff
@@ -1130,10 +1144,10 @@ class CosineRestartParamScheduler(_ParamScheduler):
...
         eta_min_ratio (float, optional): The ratio of minimum parameter value
-            to the base parameter value. Either `min_lr` or `min_lr_ratio`
+            to the base parameter value. Either `eta_min` or `eta_min_ratio`
             should be specified. Default: None.
```
Comment (HAOCHENYE) [inline mmengine/optim/scheduler/param_scheduler.py:1151]:
```suggestion
            to the base parameter value. Either `eta_min` or `eta_min_ratio`
            should be specified.
```

File: mmengine/optim/scheduler/param_scheduler.py [hunk 7]
```diff
@@ -667,21 +673,29 @@ def build_iter_from_epoch(cls,
 
     def _get_value(self):
         """Compute value using chainable form of the scheduler."""
+
+        def _get_eta_min(base_value):
+            if self.eta_min_ratio is None:
+                return self.eta_min
+            return base_value * self.eta_min_ratio
```
Comment (HAOCHENYE) [inline mmengine/optim/scheduler/param_scheduler.py:680]: Since `_get_eta_min` is only called with `base_value`. We can directly calculate the minimum value before the if else, rather than define a nested function

File: mmengine/optim/scheduler/param_scheduler.py [hunk 8]
```diff
@@ -667,21 +673,29 @@ def build_iter_from_epoch(cls,
 
     def _get_value(self):
```
Comment (HAOCHENYE) [inline mmengine/optim/scheduler/param_scheduler.py:674]:
```suggestion
    def _get_value(self) -> list:
```

File: mmengine/optim/scheduler/param_scheduler.py (hunk 6 above)
Comment (cir7) [reply mmengine/optim/scheduler/param_scheduler.py:1151]: I dont understand why delete `Default: None` here

Comment (HAOCHENYE) [reply mmengine/optim/scheduler/param_scheduler.py:1151]: type hint `Optional` means a default value with None.

Comment (HAOCHENYE) [reply mmengine/optim/scheduler/param_scheduler.py:1151]: Besides, add `New in version 0.4.0`

File: mmengine/optim/scheduler/param_scheduler.py [hunk 9]
```diff
@@ -619,14 +622,18 @@ def __init__(self,
                  optimizer: Union[Optimizer, OptimWrapper],
                  param_name: str,
                  T_max: Optional[int] = None,
-                 eta_min: float = 0.,
+                 eta_min: Optional[float] = None,
+                 eta_min_ratio: Optional[float] = None,
```
Comment (zhouzaida) [inline mmengine/optim/scheduler/param_scheduler.py:626]: Adding a new parameter in the middle position will cause backward compatibility.

Comment (zhouzaida) [reply mmengine/optim/scheduler/param_scheduler.py:626]: In addition, changing the default value of `eta_min` to None also breaks backward compatibility.

File: mmengine/optim/scheduler/param_scheduler.py [hunk 10]
```diff
@@ -1130,11 +1145,11 @@ class CosineRestartParamScheduler(_ParamScheduler):
         periods (list[int]): Periods for each cosine anneling cycle.
         restart_weights (list[float]): Restart weights at each
             restart iteration. Defaults to [1].
-        eta_min (float): Minimum parameter value at the end of scheduling.
-            Defaults to None.
+        eta_min (float, optional): Minimum parameter value at the end of
+            scheduling.
```
Comment (zhouzaida) [inline mmengine/optim/scheduler/param_scheduler.py:1149]:
```suggestion
            scheduling. Defaults to None.
```

File: mmengine/optim/scheduler/param_scheduler.py [hunk 11]
```diff
@@ -1130,11 +1145,11 @@ class CosineRestartParamScheduler(_ParamScheduler):
...
         eta_min_ratio (float, optional): The ratio of minimum parameter value
-            to the base parameter value. Either `min_lr` or `min_lr_ratio`
-            should be specified. Default: None.
+            to the base parameter value. Either `eta_min` or `eta_min_ratio`
+            should be specified.
```
Comment (zhouzaida) [inline mmengine/optim/scheduler/param_scheduler.py:1152]:
```suggestion
            should be specified. Defaults to None.
```

File: mmengine/optim/scheduler/param_scheduler.py [hunk 12]
```diff
@@ -1146,10 +1149,10 @@ class CosineRestartParamScheduler(_ParamScheduler):
         restart_weights (list[float]): Restart weights at each
             restart iteration. Defaults to [1].
         eta_min (float, optional): Minimum parameter value at the end of
-            scheduling.
+            scheduling. Defaults to None.
         eta_min_ratio (float, optional): The ratio of minimum parameter value
             to the base parameter value. Either `eta_min` or `eta_min_ratio`
-            should be specified.
+            should be specified. Defaults to None.
```
Comment (HAOCHENYE) [inline mmengine/optim/scheduler/param_scheduler.py:1155]: Update the sequence of the arguments in docstring too.

Comment (cir7) [reply mmengine/optim/scheduler/param_scheduler.py:1155]: it seems that there is no arguments update in `CosineRestartParamScheduler`

**Review by zhouzaida — APPROVED** [review]

**Review by HAOCHENYE — DISMISSED** [review]
Approved!!!
image

File: mmengine/optim/scheduler/param_scheduler.py [hunk 13]
```diff
@@ -610,7 +610,10 @@ class CosineAnnealingParamScheduler(_ParamScheduler):
             epochs. Defaults to True.
         verbose (bool): Whether to print the value for each update.
             Defaults to False.
-
+        eta_min_ratio (float, optional): The ratio of the minimum parameter
```
Comment (RangiLyu) [inline mmengine/optim/scheduler/param_scheduler.py:613]: Also need to update the doc string of CosineAnnealingLR and CosineAnnealingMomentum.

File: mmengine/optim/scheduler/lr_scheduler.py [hunk 14]
```diff
@@ -82,6 +82,10 @@ class CosineAnnealingLR(LRSchedulerMixin, CosineAnnealingParamScheduler):
             epochs. Defaults to True.
         verbose (bool): Whether to print the learning rate for each update.
             Defaults to False.
+        eta_min_ratio (float, optional): The ratio of the minimum parameter
+            value to the base parameter value. Either `eta_min` or
+            `eta_min_ratio` should be specified. Defaults to None.
+            New in version 0.4.0.
```
Comment (zhouzaida) [inline mmengine/optim/scheduler/lr_scheduler.py:88]:
```suggestion
            New in version 0.3.2.
```

File: mmengine/optim/scheduler/momentum_scheduler.py [hunk 15]
```diff
@@ -112,6 +112,10 @@ class CosineAnnealingMomentum(MomentumSchedulerMixin,
             epochs. Defaults to True.
         verbose (bool): Whether to print the momentum for each update.
             Defaults to False.
+        eta_min_ratio (float, optional): The ratio of the minimum parameter
+            value to the base parameter value. Either `eta_min` or
+            `eta_min_ratio` should be specified. Defaults to None.
+            New in version 0.4.0.
```
Comment (zhouzaida) [inline mmengine/optim/scheduler/momentum_scheduler.py:118]:
```suggestion
            New in version 0.3.2.
```

File: mmengine/optim/scheduler/param_scheduler.py [hunk 16]
```diff
@@ -610,6 +610,10 @@ class CosineAnnealingParamScheduler(_ParamScheduler):
             epochs. Defaults to True.
         verbose (bool): Whether to print the value for each update.
             Defaults to False.
+        eta_min_ratio (float, optional): The ratio of the minimum parameter
+            value to the base parameter value. Either `eta_min` or
+            `eta_min_ratio` should be specified. Defaults to None.
+            New in version 0.4.0.
```
Comment (zhouzaida) [inline mmengine/optim/scheduler/param_scheduler.py:616]:
```suggestion
            New in version 0.3.2.
```

**Review by zhouzaida — APPROVED** [review]

**Review by RangiLyu — APPROVED** [review]
